versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/objtypes/GenericObjTypeMapper.java

```diff
@@ -0,0 +1,86 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.storage.common.objtypes;
+
+import static org.projectnessie.versioned.storage.common.json.ObjIdHelper.OBJ_ID_KEY;
+import static org.projectnessie.versioned.storage.common.json.ObjIdHelper.OBJ_TYPE_KEY;
+import static org.projectnessie.versioned.storage.common.json.ObjIdHelper.OBJ_VERS_KEY;
+import static org.projectnessie.versioned.storage.common.objtypes.CustomObjType.uncachedObjType;
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.google.common.annotations.VisibleForTesting;
+import jakarta.annotation.Nonnull;
+import jakarta.annotation.Nullable;
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Target;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.immutables.value.Value;
+import org.projectnessie.versioned.storage.common.persist.ObjId;
+import org.projectnessie.versioned.storage.common.persist.ObjType;
+import org.projectnessie.versioned.storage.common.persist.UpdateableObj;
+
+/** <em>Internal</em> class used to handle otherwise unhandled object types. */
+public final class GenericObjTypeMapper {
+  @Nonnull
+  public ObjType mapGenericObjType(String nameOrShortName) {
+    return genericTypes.computeIfAbsent(nameOrShortName, GenericObjTypeMapper::newGenericObjType);
+  }
+
+  @VisibleForTesting
+  public static ObjType newGenericObjType(String name) {
+    return uncachedObjType(name, name, GenericObj.class);
+  }
+
+  private final Map<String, ObjType> genericTypes = new ConcurrentHashMap<>();
+
+  @Value.Immutable
+  @Value.Style(jdkOnly = true)
+  @JsonSerialize(as = ImmutableGenericObj.class)
+  @JsonDeserialize(as = ImmutableGenericObj.class)
+  abstract static class GenericObj implements UpdateableObj {
+    @Override
+    @JsonIgnore
+    @JacksonInject(OBJ_TYPE_KEY)
+    public abstract ObjType type();
+
+    @Override
+    @JsonIgnore
+    @JacksonInject(OBJ_ID_KEY)
+    public abstract ObjId id();
+
+    @Override
+    @JsonIgnore
+    @JacksonInject(OBJ_VERS_KEY)
+    @Nullable
+    public abstract String versionToken();
+
+    @JsonAnyGetter
+    @JsonAnySetter
+    @AllowNulls
+    public abstract Map<String, Object> attributes();
+  }
```

Comment:
I'm concerned that `GenericObj` implements `UpdateableObj`. I think that if it's generic, then it should implement only `Obj`. If the original object was updateable, I'd expect the version token to appear in the `attributes` property bag.

I understand that you might have done this because the version token is an injectable value, but there is a way to capture the injected value and include it in the property bag:

```suggestion
  @Value.Immutable
  @Value.Style(jdkOnly = true)
  abstract static class GenericObj implements Obj {
    @Override
    @JsonIgnore
    public abstract ObjType type();

    @Override
    @JsonIgnore
    public abstract ObjId id();

    @JsonAnyGetter
    @AllowNulls
    public abstract Map<String, Object> attributes();

    @JsonCreator
    static GenericObj create(
        @JacksonInject(OBJ_TYPE_KEY) ObjType objType,
        @JacksonInject(OBJ_ID_KEY) ObjId id,
        @JacksonInject(OBJ_VERS_KEY) String versionToken,
        @JsonAnySetter Map<String, Object> attributes) {
      ImmutableGenericObj.Builder builder =
          ImmutableGenericObj.builder().type(objType).id(id).attributes(attributes);
      if (versionToken != null) {
        builder.putAttributes("versionToken", versionToken);
      }
      return builder.build();
    }
  }
```

Reply:
It must implmenent UpdateableObj so that the `Persist` implementations set the field in the database appropriately.

Reply:
Conceptually, not every generic object is updateable. What is sad is that now `UpdateableObj.versionToken()` became nullable which is confusing.

Wouldn't it be better to transfer the burden of extracting the version token to `Persist` implementations? Something like:

```java
    if (obj instanceof UpdateableObj) {
      item.put(COL_OBJ_VERS, fromS(((UpdateableObj) obj).versionToken());
    }
    if (obj instanceof GenericObj && ((GenericObj)obj).hasAttribute("versionToken")) {
      item.put(COL_OBJ_VERS, fromS(((GenericObj)obj).getAttribute("versionToken"));
    }
```

(The above can be moved to a utility class of course.)

Reply:
I know it's not great. Happy to open a follow-up issue for this.

Reply:
OK, approving then. I can look into it.